halo2_proofs/src/plonk/prover.rs

```diff
@@ -127,209 +122,272 @@ pub fn create_proof<
+                return Err(Error::not_enough_rows_available(self.k));
+            }
 
-                    Ok(())
-                }
+            *self
+                .advice
+                .get_mut(column.index())
+                .and_then(|v| v.get_mut(row))
+                .ok_or(Error::BoundsFailure)? = to()?.into();
 
-                fn assign_fixed<V, VR, A, AR>(
-                    &mut self,
-                    _: A,
-                    _: Column<Fixed>,
-                    _: usize,
-                    _: V,
-                ) -> Result<(), Error>
-                where
-                    V: FnOnce() -> Result<VR, Error>,
-                    VR: Into<Assigned<F>>,
-                    A: FnOnce() -> AR,
-                    AR: Into<String>,
-                {
-                    // We only care about advice columns here
+            Ok(())
+        }
 
-                    Ok(())
-                }
+        fn assign_fixed<V, VR, A, AR>(
+            &mut self,
+            _: A,
+            _: Column<Fixed>,
+            _: usize,
+            _: V,
+        ) -> Result<(), Error>
+        where
+            V: FnOnce() -> Result<VR, Error>,
+            VR: Into<Assigned<F>>,
+            A: FnOnce() -> AR,
+            AR: Into<String>,
+        {
+            // We only care about advice columns here
+
+            Ok(())
+        }
 
-                fn copy(
-                    &mut self,
-                    _: Column<Any>,
-                    _: usize,
-                    _: Column<Any>,
-                    _: usize,
-                ) -> Result<(), Error> {
-                    // We only care about advice columns here
+        fn copy(
+            &mut self,
+            _: Column<Any>,
+            _: usize,
+            _: Column<Any>,
+            _: usize,
+        ) -> Result<(), Error> {
+            // We only care about advice columns here
 
-                    Ok(())
-                }
+            Ok(())
+        }
 
-                fn fill_from_row(
-                    &mut self,
-                    _: Column<Fixed>,
-                    _: usize,
-                    _: Option<Assigned<F>>,
-                ) -> Result<(), Error> {
-                    Ok(())
-                }
+        fn fill_from_row(
+            &mut self,
+            _: Column<Fixed>,
+            _: usize,
+            _: Option<Assigned<F>>,
+        ) -> Result<(), Error> {
+            Ok(())
+        }
 
-                fn push_namespace<NR, N>(&mut self, _: N)
-                where
-                    NR: Into<String>,
-                    N: FnOnce() -> NR,
-                {
-                    // Do nothing; we don't care about namespaces in this context.
+        fn get_challenge(&mut self, challenge: Challenge) -> Option<F> {
+            if self.current_phase > challenge.phase() {
+                Some(self.challenges[challenge.index()])
+            } else {
+                None
+            }
```

Comment:
Updated in [`ecd7081`](https://github.com/zcash/halo2/pull/593/commits/ecd70819c571d44a960f29969c9d751811d91263).

Reply:
I suggest also adding Expression::Postprocess to ensure we can apply any black-box computations to challenges.

It is partially implemented here: https://github.com/levs57/halo2 , though it probably should be refactored a bit, because right now the fact that Expression:Postprocess is formed only from Expression::Constant, Expression::Challenge or Expression::Postprocess is checked in runtime, while we could achieve the same using wrapper struct PublicExpression with private field Expression.

Usage example here: https://github.com/levs57/halo2-liam-eagen-msm/blob/master/src/testing_stuff/challenge_postprocessing_test.rs